google/cloud/firestore_v1/bulk_writer.py

```diff
@@ -0,0 +1,537 @@
+    of this, there is no companion `AsyncBulkWriter` class, as is usually seen
+    with other utility classes.
+
+    Usage:
+
+        # Instantiate the BulkWriter. This works from either `Client` or
+        # `AsyncClient`.
+        db = firestore.Client()
+        bulk_writer = db.bulk_writer()
+
+        # Attach an optional success listener. This will be called once per
+        # document.
+        bulk_writer.on_write_result(
+            lambda reference, result: print(f'Saved {reference._document_path}')
+        )
+
+        # Queue an arbitrary amount of write operations.
+        # `my_new_records` is a list of (DocumentReference, dict,) tuple-pairs
+        # that you supply.
+        reference: DocumentReference
+        data: dict
+        for reference, data in my_new_records:
+            bulk_writer.create(reference, data)
+
+        # Block until all pooled writes are complete.
+        bulk_writer.flush()
+
+    Args:
+        client(:class:`~google.cloud.firestore_v1.client.Client`):
+            The client that created this BulkWriter.
+    """
+
+    batch_size: int = 20
+
+    def __init__(
+        self,
+        client: Optional[BaseClient] = None,
+        options: Optional["BulkWriterOptions"] = None,
+    ):
+        self._client = client
+        self._options = options or BulkWriterOptions()
+        self._send_mode = self._options.mode
+        # Redundantly set this variable for IDE type hints
+        self._batch: BulkWriteBatch = self._reset_batch()
+        self._queued_batches = collections.deque([])
+        self._is_open: bool = True
+        self._is_sending: bool = False
+
+        self._success_callback: Optional[
+            Callable[[BaseDocumentReference, WriteResult, "BulkWriter"], NoReturn]
+        ] = None
+        self._batch_callback: Optional[
+            Callable[[BulkWriteBatch, BatchWriteResponse, "BulkWriter"], NoReturn]
+        ] = None
+        self._error_callback: Optional[Callable] = None
+
+        self._in_flight_documents: int = 0
+        self._rate_limiter = RateLimiter(
+            initial_tokens=self._options.initial_ops_per_second,
+            global_max_tokens=self._options.max_ops_per_second,
+        )
+
+        # Keep track of progress as batches and write operations are completed
+        self._total_batches_sent: int = 0
+        self._total_write_operations: int = 0
+
+        self._ensure_executor()
+
+    def _reset_batch(self) -> BulkWriteBatch:
+        self._batch = BulkWriteBatch(self._client)
+        return self._batch
+
+    def _ensure_executor(self):
+        self._executor = (
+            getattr(self, "_executor", None) or self._instantiate_executor()
+        )
+
+    def _ensure_sending(self):
+        if not self._is_sending:
+            self._ensure_executor()
+            self._send_until_queue_is_empty()
+
+    def _instantiate_executor(self):
+        return concurrent.futures.ThreadPoolExecutor()
```

Comment:
This serves as a customization hook, should end-users desire to tinker with the parallelization strategy.